src/main/clojure/jarl/builtins/comparison.clj

```diff
@@ -0,0 +1,116 @@
+(ns jarl.builtins.comparison
+  (:require [jarl.builtins.utils :refer [type-sort-order]]))
+
+; From the OPA go docs on ast.Compare:
+;
+; If a is less than b, the return value is negative. If a is greater than b,
+; the return value is positive. If a is equal to b, the return value is zero.
+;
+; Different types are never equal to each other. For comparison purposes, types
+; are sorted as follows:
+;
+; nil < Null < Boolean < Number < String < Var < Ref < Array < Object < Set <
+; ArrayComprehension < ObjectComprehension < SetComprehension < Expr < SomeDecl
+; < With < Body < Rule < Import < Package < Module.
+
+(declare rego-compare)
+
+(defn- first-diff-pair [a b]
+  (let [pairs (partition 2 (interleave a b))
+        diff-pairs (filter #(not= (first %) (second %)) pairs)]
+    (first diff-pairs)))
+
+; Arrays and Refs are equal if and only if both a and b have the same length
+; and all corresponding elements are equal. If one element is not equal, the
+; return value is the same as for the first differing element. If all elements
+; are equal but a and b have different lengths, the shorter is considered less
+; than the other.
+(defn- vector-compare [a b]
+  (let [ca (count a)
+        cb (count b)]
+    (if (= ca cb)
+      (let [[x y] (first-diff-pair a b)]
+        (rego-compare x y))
```

Comment:
This will just be a `nil`/`nil` comparison if `a` and `b` are equal?

Reply:
Yeah, non of these methods are really meant to be used in isolation, i.e. by other callers than `rego-compare`.. which already handles the `(= a b)` case:
```clojure
(defn rego-compare [a b]
  (cond
    (= a b) 0
```
I guess it would make sense to use `defn-`  for these to better demonstrate that. Will fix.

Reply:
Oh, they were already, it was just missing for `set-compare`.